rust/kernel/platdev.rs

```diff
@@ -49,13 +51,20 @@ impl Registration {
             return Err(Error::EINVAL);
         }
         this.pdrv.driver.name = name.as_ptr() as *const c_types::c_char;
+        if let Some(of_match_tbl) = of_match_tbl {
+            this.pdrv.driver.of_match_table = of_match_tbl.into_table_ptr();
+        }
         this.pdrv.probe = Some(probe_callback);
         this.pdrv.remove = Some(remove_callback);
         // SAFETY:
         //   - `this.pdrv` lives at least until the call to `platform_driver_unregister()` returns.
         //   - `name` pointer has static lifetime.
         //   - `module.0` lives at least as long as the module.
         //   - `probe()` and `remove()` are static functions.
+        //   - `of_match_tbl` is either:
+        //      - a raw pointer created from a heap structure which is never deallocated,
+        //        therefore it lives at least as long as the module, or
+        //      - null.
```

Comment:
I think we also need to support `'static` pointers, which would be created by `const` functions. Unless, of course, there is a technical reason that prevents us from doing so.

Reply:
I agree 100% that the structure should be `'static`, `const`, and created at build time - I even indicated that in the PR's first commit message.

If we can go that route, your remaining concerns disappear, correct?
- match table would be a`'static` pointer
- we would not have to use a `Box` (ugh that looked bad - I know !)
- length checks at runtime (these are perfect candidates for build time checks)

In this PR's v2, I can get rid of the ugly `Box` issue, and make sure that it gets cleaned up properly, without resulting in a memory leak.

The problem is - I have no idea (yet!) how to create the table statically at build time.

Could I convince you to defer this issue until a later PR?
Can't get the perfect get in the way of the good?

Reply:
PS I tried using const generics when creating the match table structure. That's probably one of the things we'll need to create the table at build time? I ran into internal compiler errors, which is kind of an indication we're in uncharted territory here...

Reply:
Yes, if we don't have the box issue, I think we could get this is and improve on it later.